File: 208-implement-trie-prefix-tree/208-implement-trie-prefix-tree.py
class TrieNode:
    def __init__(self, char=""):
        self.char = char
        self.isEnd = False
        self.children = {}
        
class Trie:

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        current = self.root
        for char in word:
            if char in current.children:
                current = current.children[char]
            else:
                newNode = TrieNode(char)
                current.children[char] = newNode
                current = newNode
        current.isEnd = True

    def search(self, word: str, isPrefix=False) -> bool:
        current = self.root
        for char in word:
            if char not in current.children:
                return False
            current = current.children[char]
        return current.isEnd or isPrefix

    def startsWith(self, prefix: str) -> bool:
        return self.search(prefix, True)


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: 208-implement-trie-prefix-tree/208-implement-trie-prefix-tree.cpp
class Trie {
    struct TrieNode {
        bool isEnd;
        unordered_map<char, TrieNode*> children;
    };
    
    TrieNode* root;
    public:
    Trie() {
        root = new TrieNode();
    }

    void insert(string word) {
        TrieNode* cur = root;
        for (auto ch: word) {
            if (cur->children.count(ch))
                cur = cur->children[ch];
            else {
                TrieNode* newNode = new TrieNode();
                cur->children[ch] = newNode;
                cur = newNode;
            }
        }
        cur->isEnd = true;
    }

    bool search(string word) {
       TrieNode* cur = root;
        for (auto ch: word) {
            if (!cur->children.count(ch))
                return false;
            cur = cur->children[ch];
        }
        return cur->isEnd;
    }

    bool startsWith(string prefix) {
       TrieNode* cur = root;
        for (auto ch: prefix) {
            if (!cur->children.count(ch))
                return false;
            cur = cur->children[ch];
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */
